### DISCUSSION TITLE
Supporting Generator with parameters

### DISCUSSION THREAD

**Discussion by mmarcinmichal** [coady/multimethod#57] [Ideas]
Hi,
it will be greate if the package supports Generator with types. Currently, it is working only on typing.Generator, i.e.

```python
class Foo2:
    @multimethod
    def bar(self, x1: str):
        print("str " + x1)

    @bar.register
    def _(self, x1: typing.Generator):
        print("Generator ")


tt = Foo2()
tt.bar("ala")
my_gen = (i for i in range(10))
tt.bar(my_gen)
```

When I try specidied generator type, i.e.

```python
class Foo2:
    @multimethod
    def bar(self, x1: str):
        print("str " + x1)

    @bar.register
    def _(self, x1: typing.Generator[int, None, None]):
        print("Generator ")


tt = Foo2()
tt.bar("ala")
my_gen = (i for i in range(10))
tt.bar(my_gen)
```

```
raise DispatchError(msg, types, keys)
multimethod.DispatchError: ('bar: 0 methods found', (<class '__main__.Foo2'>, <class 'generator'>), [])
```

Best,

**Comment by coady** [comment]
I don't see any way of doing it though, for iterators in general. `Sequence[...]` checks the first value - which is already inaccurate if the sequence has mixed types - but it simply can't check `Iterator[...]` without altering it. Even if it cached the first value and constructed a new iterator, the function would receive a different object.

It could partially support `Iterator[...]` by only checking for `Iterator`, but I think that would be surprising too. I'm not sure that's better, unless most users think it's better.
